hokusai/services/command_runner.py

```diff
@@ -1,71 +1,177 @@
-import os
-import re
+import copy
 import json
 import pipes
+import re
+
+import pdb
 
+from hokusai.lib.common import (
+  k8s_uuid, returncode, shout, user, validate_key_value
+)
 from hokusai.lib.config import config
-from hokusai.lib.common import shout, returncode, k8s_uuid
+from hokusai.lib.exceptions import HokusaiError
 from hokusai.services.ecr import ECR
 from hokusai.services.kubectl import Kubectl
-from hokusai.lib.exceptions import HokusaiError
 
 class CommandRunner:
   def __init__(self, context, namespace=None):
     self.context = context
     self.kctl = Kubectl(self.context, namespace=namespace)
     self.ecr = ECR()
+    self.pod_name = self._name()
+    self.container_name = self.pod_name
 
-  def run(self, tag_or_digest, cmd, tty=None, env=(), constraint=()):
-    if os.environ.get('USER') is not None:
-      # The regex used for the validation of name is '[a-z0-9]([-a-z0-9]*[a-z0-9])?'
-      user = re.sub("[^0-9a-z]+", "-", os.environ.get('USER').lower())
-      uuid = "%s-%s" % (user, k8s_uuid())
-    else:
-      uuid = k8s_uuid()
+  def _name(self):
+    ''' generate name for pod and container '''
+    name = '-'.join(
+      filter(
+        None,
+        [
+          f'{config.project_name}-hokusai-run',
+          user(),
+          k8s_uuid()
+        ]
+      )
+    )
+    return name
 
-    name = "%s-hokusai-run-%s" % (config.project_name, uuid)
-    separator = "@" if ":" in tag_or_digest else ":"
-    image_name = "%s%s%s" % (self.ecr.project_repo, separator, tag_or_digest)
+  def _image_name(self, tag_or_digest):
+    ''' generate docker image name '''
+    separator = '@' if ':' in tag_or_digest else ':'
+    image_name = f'{self.ecr.project_repo}{separator}{tag_or_digest}'
+    return image_name
 
-    container = {
-      "args": cmd.split(' '),
-      "name": name,
-      "image": image_name,
-      "imagePullPolicy": "Always",
-      'envFrom': [{'configMapRef': {'name': "%s-environment" % config.project_name}}]
-    }
+  def _set_env(self, container_spec, env):
+    ''' set env field of given container spec '''
+    spec = copy.deepcopy(container_spec)
```

Comment:
Type hints would be useful here... it looks like `container_spec` is a dict, in which case I'm wondering why you are deepcopying it. Can you not just mutate it and return the mutated dict.

Reply:
> Can you not just mutate it...

Yeah, I prefer not to mutate args, when at all possible. The `deepcopy` is overhead, but the dict is not too large, so shouldn't hurt performance too much.